workflow-bot-app/src/main/java/com/symphony/bdk/workflow/engine/WorkflowDirectGraphBuilder.java

```diff
@@ -0,0 +1,229 @@
+        } else if (activity.getActivity().getElseCondition() != null) {
+          throw new InvalidActivityException(workflow.getId(),
+              "Expecting \"if\" keyword to open a new conditional branching, got \"else\"");
+        }
+      }
+    }
+  }
+
+  private void computeEvents(int activityIndex, String activityId, RelationalEvents onEvents, List<Activity> activities,
+      WorkflowDirectGraph directGraph) {
+    for (Event event : onEvents.getEvents()) {
+      String eventNodeId = "";
+      if (isTimerFiredEvent(event)) {
+        eventNodeId = eventMapper.toTimerFiredEventName(event.getTimerFired());
+        computeActivity(activityIndex, activities, eventNodeId, event, onEvents.isExclusive(), directGraph);
+        directGraph.registerToDictionary(eventNodeId,
+            new WorkflowNode().id(eventNodeId).event(event).elementType(WorkflowNodeType.TIMER_FIRED_EVENT));
+      } else {
+        Optional<String> signalName = eventMapper.toSignalName(event, workflow);
+        if (signalName.isPresent()) {
+          eventNodeId = signalName.get();
+          computeActivity(activityIndex, activities, eventNodeId, event, onEvents.isExclusive(), directGraph);
+          computeSignal(directGraph, event, eventNodeId, activityIndex, activityId, activities);
+        } else if (event.getActivityExpired() != null) {
+          eventNodeId = computeExpiredActivity(event, activityId, directGraph);
+        } else if (event.getActivityFailed() != null) {
+          eventNodeId = event.getActivityFailed().getActivityId();
+          validateNodeId(eventNodeId, activityId);
+          directGraph.readWorkflowNode(activityId).setElementType(WorkflowNodeType.ACTIVITY_FAILED_EVENT);
+        } else if (event.getActivityCompleted() != null) {
+          eventNodeId = event.getActivityCompleted().getActivityId();
+          validateNodeId(eventNodeId, activityId);
+          directGraph.readWorkflowNode(eventNodeId).setElementType(WorkflowNodeType.ACTIVITY_COMPLETED_EVENT);
+          BaseActivity currentActivity = activities.get(activityIndex).getActivity();
+          Optional<String> condition = retrieveCondition(event.getActivityCompleted(), currentActivity, directGraph);
+          String finalNodeId = eventNodeId;
+          condition.ifPresent(c -> directGraph.readWorkflowNode(activityId).addIfCondition(finalNodeId, c));
+        }
+      }
+      directGraph.addChildTo(eventNodeId)
+          .gateway(onEvents.isExclusive() ? Gateway.EVENT_BASED : Gateway.PARALLEL)
+          .addChild(activityId);
+      directGraph.addParent(activityId, eventNodeId);
+    }
+  }
+
+  private Optional<String> retrieveCondition(ActivityCompletedEvent event, BaseActivity activity,
+      WorkflowDirectGraph directGraph) {
+    return event.getIfCondition() != null ? Optional.of(event.getIfCondition())
+        : Optional.ofNullable(activity.getIfCondition());
+  }
+
+  private void validateNodeId(String currentNodeId, String activityId) {
```

Comment:
I don't think this validation is enough.. When an activity act2 has an on:event like on-activity-completed referencing  an activity act1, we don't only check that the activity act1 exists, but it should also be prior to act2 in the workflow. In fact an on-activity-completed or on-activity-expired.. cannot reference an activity coming next.

Reply:
I just saw this method validateActivity(), maybe it needs to be enriched with more constraint checks?